Swap to using environment variables for hard coded certs

He currently have hard coded cert names in our broker

https://github.com/NOAA-OWP/wres/blob/bc993975a9ce41be6d0dc9eb54daa36d8ecb4e1c/wres-broker/nonsrc/rabbitmq.conf#L10-L23

Also there is this hard coded pass for our trust store. Which doesn't really do anything but for craftmanship should be extracted as well ( think it also shows up in a test file as well) 

https://github.com/NOAA-OWP/wres/blob/bc993975a9ce41be6d0dc9eb54daa36d8ecb4e1c/wres-system/src/wres/system/SSLStuffThatTrustsOneCertificate.java#L111-L115

https://github.com/NOAA-OWP/wres/blob/bc993975a9ce41be6d0dc9eb54daa36d8ecb4e1c/wres-messages/src/main/java/wres/messages/BrokerHelper.java#L171-L172

After custom building my compose for the entry machine, I deployed without updating the `.env` and received the error I expected in the tasker and worker; for example, in the tasker:

```
tasker_1        | Exception in thread "main" java.lang.ExceptionInInitializerError
tasker_1        |       at wres.tasker.Tasker.main(Tasker.java:134)
tasker_1        | Caused by: java.lang.IllegalStateException: Both wres.trustStore and wres.trustStorePassword must be specified to initialize the trust manager.
tasker_1        |       at wres.messages.BrokerHelper.getDefaultTrustManager(BrokerHelper.java:165)
tasker_1        |       at wres.messages.BrokerHelper.getSSLContextWithClientCertificate(BrokerHelper.java:355)
tasker_1        |       at wres.tasker.WresJob.initializeBrokerConnectionFactory(WresJob.java:1006)
tasker_1        |       at wres.tasker.WresJob.<clinit>(WresJob.java:144)
tasker_1        |       ... 1 more
```

I then added the password to the .env, `WRES_TRUST_STORE_PASSWORD`, and it worked.  A smoke test passed. 

The events broker change is also clear cut, so I'm working on that one next. Per the conversation with James, the change to `SSLStuffThatTrustsOneCertificate` may be a bit broader than the other changes, so I'll do that one last.  

Oh, and Evan mentioned a unit test that may be impacted by the "changeit" changes, so I'll look for that.

Hank

So I ran a couple more evaluations, including one using WRDS, and they all succeeded. 

However, when I brought down the service, commented out the tasker and worker password env variables, and brought it back up, I got NPEs:

```
java.lang.IllegalStateException: WRES expected to be able to read and decrypt the file '/wres_secrets/wres-tasker_client_private_key_and_x509_cert.p12'.
        at wres.messages.BrokerHelper.getSSLContextWithClientCertificate(BrokerHelper.java:332)
        at wres.tasker.WresJob.initializeBrokerConnectionFactory(WresJob.java:1014)
        at wres.tasker.WresJob.<clinit>(WresJob.java:144)
        at wres.tasker.Tasker.main(Tasker.java:138)
Caused by: java.security.UnrecoverableKeyException: Get Key failed: Cannot read the array length because "password" is null
        at java.base/sun.security.pkcs12.PKCS12KeyStore.engineGetKey(PKCS12KeyStore.java:450)
        at java.base/sun.security.util.KeyStoreDelegator.engineGetKey(KeyStoreDelegator.java:91)
        at java.base/java.security.KeyStore.getKey(KeyStore.java:1050)
        at java.base/sun.security.ssl.SunX509KeyManagerImpl.<init>(SunX509KeyManagerImpl.java:141)
        at java.base/sun.security.ssl.KeyManagerFactoryImpl$SunX509.engineInit(KeyManagerFactoryImpl.java:64)
        at java.base/javax.net.ssl.KeyManagerFactory.init(KeyManagerFactory.java:275)
        at wres.messages.BrokerHelper.getSSLContextWithClientCertificate(BrokerHelper.java:328)
        ... 3 common frames omitted
Caused by: java.lang.NullPointerException: Cannot read the array length because "password" is null
        at java.base/sun.security.pkcs12.PKCS12KeyStore$RetryWithZero.run(PKCS12KeyStore.java:259)
        at java.base/sun.security.pkcs12.PKCS12KeyStore.engineGetKey(PKCS12KeyStore.java:361)
        ... 9 common frames omitted
Exception in thread "main" java.lang.ExceptionInInitializerError
        at wres.tasker.Tasker.main(Tasker.java:138)
Caused by: java.lang.IllegalStateException: WRES expected to be able to read and decrypt the file '/wres_secrets/wres-tasker_client_private_key_and_x509_cert.p12'.
        at wres.messages.BrokerHelper.getSSLContextWithClientCertificate(BrokerHelper.java:332)
        at wres.tasker.WresJob.initializeBrokerConnectionFactory(WresJob.java:1014)
        at wres.tasker.WresJob.<clinit>(WresJob.java:144)
        ... 1 more
Caused by: java.security.UnrecoverableKeyException: Get Key failed: Cannot read the array length because "password" is null
        at java.base/sun.security.pkcs12.PKCS12KeyStore.engineGetKey(PKCS12KeyStore.java:450)
        at java.base/sun.security.util.KeyStoreDelegator.engineGetKey(KeyStoreDelegator.java:91)
        at java.base/java.security.KeyStore.getKey(KeyStore.java:1050)
        at java.base/sun.security.ssl.SunX509KeyManagerImpl.<init>(SunX509KeyManagerImpl.java:141)
        at java.base/sun.security.ssl.KeyManagerFactoryImpl$SunX509.engineInit(KeyManagerFactoryImpl.java:64)
        at java.base/javax.net.ssl.KeyManagerFactory.init(KeyManagerFactory.java:275)
        at wres.messages.BrokerHelper.getSSLContextWithClientCertificate(BrokerHelper.java:328)
        ... 3 more
Caused by: java.lang.NullPointerException: Cannot read the array length because "password" is null
        at java.base/sun.security.pkcs12.PKCS12KeyStore$RetryWithZero.run(PKCS12KeyStore.java:259)
        at java.base/sun.security.pkcs12.PKCS12KeyStore.engineGetKey(PKCS12KeyStore.java:361)
        ... 9 more
```

The reason is because, `keyManagerFactory.init( keyStore, keyPassphrase );`, cannot handle a null passphrase.

Will look into that after a meeting,

Hank

The Javadoc doesn't talk about null or empty arrays, so it wasn't helpful. I made the change to use an empty array if no password is provided and testing it results in the expected log message in the tasker:

```
Exception in thread "main" java.lang.ExceptionInInitializerError
        at wres.tasker.Tasker.main(Tasker.java:138)
Caused by: java.lang.IllegalStateException: WRES expected to find a file '/wres_secrets/wres-tasker_client_private_key_and_x509_cert.p12' with PKCS#12 format, with both a client certificate AND the private key inside, used to authenticate to the broker.
        at wres.messages.BrokerHelper.getSSLContextWithClientCertificate(BrokerHelper.java:305)
        at wres.tasker.WresJob.initializeBrokerConnectionFactory(WresJob.java:1014)
        at wres.tasker.WresJob.<clinit>(WresJob.java:144)
        ... 1 more
Caused by: java.io.IOException: keystore password was incorrect
        at java.base/sun.security.pkcs12.PKCS12KeyStore.engineLoad(PKCS12KeyStore.java:2159)
        at java.base/sun.security.util.KeyStoreDelegator.engineLoad(KeyStoreDelegator.java:221)
        at java.base/java.security.KeyStore.load(KeyStore.java:1473)
        at wres.messages.BrokerHelper.getSSLContextWithClientCertificate(BrokerHelper.java:301)
        ... 3 more
```

and in the workers:

```
Exception in thread "main" java.lang.IllegalStateException: WRES expected to find a file '/wres_secrets/wres-worker_client_private_key_and_x509_cert.p12' with PKCS#12 format, with both a client certificate AND the private key inside, used to authenticate to the broker.
        at wres.messages.BrokerHelper.getSSLContextWithClientCertificate(BrokerHelper.java:305)
        at wres.worker.Worker.createConnectionFactory(Worker.java:205)
        at wres.worker.Worker.main(Worker.java:105)
Caused by: java.io.IOException: keystore password was incorrect
        at java.base/sun.security.pkcs12.PKCS12KeyStore.engineLoad(PKCS12KeyStore.java:2159)
        at java.base/sun.security.util.KeyStoreDelegator.engineLoad(KeyStoreDelegator.java:221)
        at java.base/java.security.KeyStore.load(KeyStore.java:1473)
        at wres.messages.BrokerHelper.getSSLContextWithClientCertificate(BrokerHelper.java:301)
        ... 2 more
```

Cool. I'm going to bring back the environment variables. cycle the -dev COWRES, and then focus on my 3pm meeting. That's it for me today for this ticket. I'll complete testing tomorrow and get the code pushed.

Hank

Unit tests are failing for me:

```
Caused by: java.lang.IllegalStateException: The system property wres.taskerPathToClientP12Bundle is not specified. It must be provided and non-empty.
	at wres.tasker.WresJob.initializeBrokerConnectionFactory(WresJob.java:1000)
	at wres.tasker.WresJob.<clinit>(WresJob.java:144)
	... 59 more
```

--- a/wres-tasker/test/wres/tasker/WresJobTest.java
+++ b/wres-tasker/test/wres/tasker/WresJobTest.java
@@ -64,24 +64,24 @@ public class WresJobTest
     private static final String SERVER_KEY_FILE_NAME = "wres-broker-localhost_server_private_rsa_key.pem";
     private static final String TRUST_STORE_FILE_NAME = "trustedCertificates-localhost.jks";
     private static final String[] APPROVED_CIPHERS = {
-            "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256",
-            "TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384",
-            "TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA",
-            "TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA",
-            "TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256",
-            "TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384",
-            "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256",
-            "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384",
-            "TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA",
-            "TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA",
-            "TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256",
-            "TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384",
-            "TLS_DHE_RSA_WITH_AES_128_GCM_SHA256",
-            "TLS_DHE_RSA_WITH_AES_256_GCM_SHA384",
-            "TLS_DHE_RSA_WITH_AES_128_CBC_SHA",
-            "TLS_DHE_RSA_WITH_AES_256_CBC_SHA",
-            "TLS_DHE_RSA_WITH_AES_128_CBC_SHA256",
-            "TLS_DHE_RSA_WITH_AES_256_CBC_SHA256"
+                                                       "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256",
+                                                       "TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384",
+                                                       "TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA",
+                                                       "TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA",
+                                                       "TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256",
+                                                       "TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384",
+                                                       "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256",
+                                                       "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384",
+                                                       "TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA",
+                                                       "TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA",
+                                                       "TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256",
+                                                       "TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384",
+                                                       "TLS_DHE_RSA_WITH_AES_128_GCM_SHA256",
+                                                       "TLS_DHE_RSA_WITH_AES_256_GCM_SHA384",
+                                                       "TLS_DHE_RSA_WITH_AES_128_CBC_SHA",
+                                                       "TLS_DHE_RSA_WITH_AES_256_CBC_SHA",
+                                                       "TLS_DHE_RSA_WITH_AES_128_CBC_SHA256",
+                                                       "TLS_DHE_RSA_WITH_AES_256_CBC_SHA256"
     };
 
     private static final String FAKE_DECLARATION = """
@@ -108,9 +108,9 @@ public class WresJobTest
 
         // Create a certificate
         X500Name subject = new X500NameBuilder( BCStyle.INSTANCE )
-                .addRDN( BCStyle.CN, "localhost" )
-                .addRDN( BCStyle.O, "WRES" )
-                .build();
+                                                                  .addRDN( BCStyle.CN, "localhost" )
+                                                                  .addRDN( BCStyle.O, "WRES" )
+                                                                  .build();
 
         Instant now = Instant.now();
         Instant aLittleWhileAgo = now.minus( 10_000, ChronoUnit.SECONDS );
@@ -152,7 +152,7 @@ public class WresJobTest
                                          extendedKeyUsage );
 
         ContentSigner contentSigner = new JcaContentSignerBuilder( "SHA256withRSA" )
-                .build( keyPair.getPrivate() );
+                                                                                    .build( keyPair.getPrivate() );
         X509CertificateHolder certificateHolder = certificateBuilder.build( contentSigner );
 
         JcaX509CertificateConverter certificateConverter = new JcaX509CertificateConverter();
@@ -167,7 +167,7 @@ public class WresJobTest
 
         KeyStore.PrivateKeyEntry privateKeyEntry = new KeyStore.PrivateKeyEntry( keyPair.getPrivate(),
                                                                                  new Certificate[] {
-                                                                                         wowACertificate } );
+                                                                                                     wowACertificate } );
         p12KeyStore.setEntry( "privateKeyAlias",
                               privateKeyEntry,
                               new KeyStore.PasswordProtection( passphrase.toCharArray() ) );
@@ -240,6 +240,12 @@ public class WresJobTest
 
         System.setProperty( BrokerManagerHelper.WRES_MONITOR_PASSWORD_SYSTEM_PROPERTY_NAME,
                             "test" );
+
+        //System properties for the client P12 file handed off to the broker. These settings are based
+        //on how the file was written, above. 
+        System.setProperty( Tasker.PATH_TO_CLIENT_P12_PNAME,
+                            WresJobTest.tempDir.toString() + "/" + WresJobTest.P12_FILE_NAME );
+        System.setProperty( Tasker.PASSWORD_TO_CLIENT_P12, passphrase );
     }
 
     @Test
@@ -247,7 +253,12 @@ public class WresJobTest
     {
         System.setProperty( "wres.secrets_dir", WresJobTest.tempDir.toString() );
         WresJob wresJob = new WresJob();
-        try ( Response response = wresJob.postWresJob( "fake", "hank", null, "boogaflickle", false, false,
+        try ( Response response = wresJob.postWresJob( "fake",
+                                                       "hank",
+                                                       null,
+                                                       "boogaflickle",
+                                                       false,
+                                                       false,
                                                        Collections.emptyList() ) )
         {
             assertEquals( "Expected a 400 bad request.", 400, response.getStatus() );
@@ -318,7 +329,7 @@ public class WresJobTest
         System.setProperty( "wres.secrets_dir", WresJobTest.tempDir.toString() );
         WresJob wresJob = new WresJob();
         try ( Response response =
-                      wresJob.postWresJob( FAKE_DECLARATION, null, null, null, false, false, Collections.emptyList() ) )
+                wresJob.postWresJob( FAKE_DECLARATION, null, null, null, false, false, Collections.emptyList() ) )
         {
             assertEquals( "Expected a 500 Internal Server Error.", 500, response.getStatus() );
         }
@@ -329,7 +340,7 @@ public class WresJobTest
         {
             String[] additionalArguments = new String[] { "host", "8000", "name" };
             List<String> argsList = Arrays.asList ( additionalArguments );
-
+    
             EmbeddedBroker embeddedBroker = new EmbeddedBroker();
             embeddedBroker.start();
             WresJob wresJob = new WresJob();
@@ -360,3 +371,4 @@ public class WresJobTest
         Files.deleteIfExists( WresJobTest.tempDir );
     }
 }
+

